stino-x/priority-pro
# Fix Task Fetching & Rendering in Table Component

Description: This pull request addresses an issue in the Tasks Table component where tasks were not correctly fetched and mapped, leading to runtime errors. The key changes involve ensuring the correct structure is returned from the getTasks API call and implementing proper error handling and mapping for the tasks data.

Issue Description:
The following error was occurring in the Tasks Table component:

javascript

Error fetching tasks: TypeError: tasks.map is not a function
This was caused by an incorrect assumption that the getTasks function returned a list (array) of tasks directly, while it actually returned an object with a documents array.

Changes Made:
Task Fetching Fix:

In the getTasks function, we now correctly handle the response object by extracting the documents field, which contains the array of tasks.

Added a safeguard to check if the documents field exists and is an array. If not, the function returns an empty array, avoiding runtime errors.

Original Code:

js
const tasks = await getTasks();
const tasksList = tasks.map(task => ({
  task: task.id,
  ...
}));
Updated Code:

ts
const tasksResponse = await getTasks();
const tasks = tasksResponse?.documents || []; // Handle tasks in 'documents' field

if (!Array.isArray(tasks)) {
  return []; // Safeguard to return empty array if tasks is not valid
}

const tasksList = tasks.map(task => ({
  task: task.id,
  ...
}));
Improved Error Handling:

Added logging in case the getTasks call does not return tasks in the expected format or the data is missing.
This ensures debugging is easier in case of future issues.
Updated getData Function:

Modified getData to handle the promise returned by getTasks and map the tasks to the correct structure expected by the table.

The new logic ensures that tasks are safely mapped even when the API structure changes or is incomplete.

Old getData Implementation:

js

export async function getData() {
  const tasks = await getTasks();
  return tasks.map(task => ...);
}
New getData Implementation:

ts

export async function getData(): Promise<Payment[]> {
  try {
    const tasksResponse = await getTasks();
    const tasks = tasksResponse?.documents || [];
    
    if (!Array.isArray(tasks)) {
      console.log('Expected tasks to be an array, but got:', typeof tasks);
      return [];
    }

    return tasks.map((task) => ({
      task: task.id,
      due_date: task.due_date,
      assignee: task.assignee,
      status: task.status,
    }));
  } catch (error) {
    console.error('Error fetching tasks:', error);
    return [];
  }
}
Testing:
Manual Testing:
Verified that tasks are successfully fetched from the database and rendered in the table.
Ensured that the component handles cases where no tasks are returned, logging an appropriate message.
Tested the flow when the getTasks function throws an error, and confirmed that the error is caught and handled gracefully.
Checklist:
 Fix issue with task data not being mapped correctly.
 Ensure proper error handling in getTasks API call.
 Improve logging for easier debugging.
 Test the component with a variety of task inputs, including empty task lists and erroneous responses.
